src/storage/src/render/upsert.rs

```diff
@@ -8,1045 +8,274 @@
+                            // Skip this command if the next one is for the same (time, key) pair
+                            if p_ts == &ts && p.keyed(&key_indices) == cmd.keyed(&key_indices) {
+                                continue;
+                            }
+                        }
 
-            match data.value {
-                Some(Ok(_value)) => {
-                    panic!("got NULL key but some value that is not an error or NULL as well");
-                }
-                Some(Err(err)) => {
-                    if let Some(old_value) = current_value {
-                        // retract old value
-                        session.give((old_value.clone(), cap.time().clone(), -1));
+                        match cmd {
+                            UpsertCommand::Insert(result) => {
+                                let new_cmd =
+                                    UpsertCommand::Insert(result.clone()).into_keyed(&key_indices);
+                                if let Some(old_cmd) = state.replace(new_cmd) {
+                                    let UpsertCommand::Insert(old_result) = old_cmd.command else {
+                                        unreachable!()
+                                    };
+                                    output_updates.push((old_result, ts.clone(), -1));
+                                }
+                                output_updates.push((result, ts, 1));
+                            }
+                            UpsertCommand::Remove(result) => {
+                                let cmd = UpsertCommand::Remove(result).into_keyed(&key_indices);
+                                if let Some(old_cmd) = state.take(&cmd) {
+                                    let UpsertCommand::Insert(old_result) = old_cmd.command else {
+                                        unreachable!()
+                                    };
+                                    output_updates.push((old_result, ts, -1));
+                                }
+                            }
+                        }
                     }
-                    let err = Err(EnvelopeError::Upsert(err).into());
-                    current_values.insert(None, err.clone());
-                    session.give((err, cap.time().clone(), 1));
-                }
-                None => {
-                    if let Some(old_value) = current_value {
-                        // retract old value
-                        session.give((old_value.clone(), cap.time().clone(), -1));
 
-                        current_values.remove(&None);
+                    // Retain batches that still have data
+                    pending_batches.retain_mut(|batch| batch.peek().is_some());
+
+                    output_handle
+                        .give_container(&output_cap, &mut output_updates)
+                        .await;
+                    if let Some(ts) = upper.as_option() {
+                        output_cap.downgrade(ts);
```

Comment:
nice use of the `output_cap` instead of holding the data caps all over the place... don't you want yo return if the upper is empty?

Reply:
The `while let Some(..)` loop will naturally terminate and everything will finish so we don't need to handle it manually